text/0000-lwc-dynamic-with-attrs.md

```diff
@@ -0,0 +1,164 @@
+---
+title: LWC Dynamic with Attrs/Props/Event listeners
+status: DRAFTED
+created_at: 2021-10-01
+updated_at: 2021-10-01
+pr: (leave this empty until the PR is created)
+---
+
+# LWC Dynamic with Attrs/Props/Event listeners
+
+## Summary
+
+This proposal augments `lwc:dynamic` to not just accept the constructor, but also
+the properties/attributes/event listeners that are passed to the component.
+This allows them to be changed dynamically.
+
+## Terminology
+*component config*: properties/attributes/event/constructor listeners on a component
+
+## Basic example
+
+```html
+// x/app.html
+<template>
+    <x-cmp lwc:dynamic={cmpConfig}></x-cmp>
+</template>
+```
+
+```js
+// x/app.js
+import { LightningElement } from 'lwc';
+
+export default class App extends LightningElement {
+    cmpConfig;
+    async loadCtor() {
+        const constructor = await import("x/child");
+
+        const attrs = { style: 'color: red;' };
+        const props = { name: 'John Foo' };
+        const eventListeners = { onchange: handleChange };
+
+        this.cmpConfig = { constructor, attrs, props, eventListeners };
+    }
+}
+```
+In the above example, `<x-cmp>` is created with the constructor `x/child`. The `<x-cmp>` element will have
+the `style` as an attribute, `name` assigned as a property and `handleChange` added as an event listener of type `change`.
+
+## Motivation
+
+`lwc:dynamic` allows component authors to lazily instantiate components. However,
+the props that we pass to the component can't be dynamic: we need to pass them
+in the template HTML itself.
+
+This is a problem when the component constructor is changed at runtime: the new constructor,
+might need a different set of properties than the earlier one.
+
+We can, of course, not pass any props and once the element renders, query the
+element and set the props programmatically. We lose reactivity this way and also force
+the component to handle `null` props which might not be desirable.
+
+This proposal allows `lwc:dynamic` to accept a *config* along with the constructor.
+
+## Detailed design
+
+`lwc:dynamic` currently accepts a `LightningElementConstructor` to be passed. With this
+proposal, we modify it to accept `LightningElementConstructor | LightningElementConfig` where
```

Comment:
Another possible approach is:
- leaving lwc:dynamic as is
- add a new prop attribute called 'lwc:dynamic-props' that takes in the prop config

It would behave similarly but also allow LWC to gate usage of dynamic props via the linter (also in the future if we decide to add attrs and listeners, we could gate those individually).